SwrveSDK/src/google/java/com/swrve/sdk/SwrvePushEngageReceiver.java

```diff
@@ -0,0 +1,113 @@
+package com.swrve.sdk;
+
+import android.app.PendingIntent;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+
+import com.swrve.sdk.gcm.SwrveGcmConstants;
+import com.swrve.sdk.gcm.SwrveGcmNotification;
+
+import org.json.JSONException;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SwrvePushEngageReceiver extends BroadcastReceiver {
+
+    private static final String LOG_TAG = "SwrveGcmEngageReceiver";
+
+    private Context context;
+    private Swrve swrve;
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        try {
+            this.context = context;
+            swrve = (Swrve) SwrveSDK.getInstance();
+            processIntent(intent);
+        } catch (Exception ex) {
+            SwrveLogger.e(LOG_TAG, "SwrvePushEngageReceiver. Error processing intent. Intent:" + intent, ex);
+        }
+    }
+
+    private void processIntent(Intent intent) {
+        if (intent != null) {
+            Bundle extras = intent.getExtras();
+            if (extras != null && !extras.isEmpty()) {
+                Bundle msg = extras.getBundle(SwrveGcmConstants.GCM_BUNDLE);
+                if (msg != null && SwrveSDK.getConfig().isPushEnabled()) {
+                    // Obtain push id
+                    Object rawId = msg.get(SwrveGcmConstants.SWRVE_TRACKING_KEY);
+                    String msgId = (rawId != null) ? rawId.toString() : null;
+                    // Only process once the message if possible
+                    if (!SwrveHelper.isNullOrEmpty(msgId)) {
+                        String eventName = "Swrve.Messages.Push-" + msgId + ".engaged";
+                        SwrveLogger.d(LOG_TAG, "GCM engaged, sending event:" + eventName);
+                        sendEngagedEvent(eventName);
+                        // Call custom listener
+                        if (swrve.pushNotificationListener != null) {
+                            swrve.pushNotificationListener.onPushNotification(msg); // todo warning that this should not take long......
```

Comment:
Need to add notes/docs that this callback is now being executed in a broadcast receiver so needs to be fast or else handled safely in another thread